### PR TITLE
DCOS-10983: Add multiple containers to the service create form

### PR DESCRIPTION
**Pull request by Poltergeist**
As requested by @orlandohohmeier here is a gif:
giphy

and here is a screen capture:

pods-multi-container mov

### CODE REVIEW

**Comment by nLight** [comment]
When I try to add Env vars I get
```
Uncaught TypeError: Cannot read property 'map' of undefined in EnvironmentFormSection.js getEnvironmentLines
```

is it related to the PR?

**Comment by Poltergeist** [comment]
Yes it is and I just fixed the issue.

**Comment by nLight** [comment]
Can't type `0` into CPUs field of a container when I want to have less then 1 CPU

File: plugins/services/src/js/components/forms/GeneralServiceFormSection.js [hunk 1]
```diff
@@ -36,6 +43,89 @@ const containerRuntimes = {
...
+          onRemoveItem={this.props.onRemoveItem} />
+      );
+    }
+
+    let {containers = []} = data;
+    return containers.map((item, index) => {
+      return (
+        <div key={index}>
+          {item.name || `container ${index + 1}`}
+          <a className="button button-primary-link"
```
Comment (nLight) [inline plugins/services/src/js/components/forms/GeneralServiceFormSection.js]: We might want to use the new delete cross icon here

File: plugins/services/src/js/components/forms/GeneralServiceFormSection.js [hunk 2]
```diff
@@ -36,6 +43,89 @@ const containerRuntimes = {
...
+  getConvertToPodAction() {
+    let {service, isEdit} = this.props;
+
+    if (isEdit || service instanceof Pod) {
+      return null;
+    }
+
+    return (
+      <p>
+        Need to run a service with multiple containers?
```
Comment (nLight) [inline plugins/services/src/js/components/forms/GeneralServiceFormSection.js]: Let's add a space between this text and the link

File: plugins/services/src/js/components/forms/GeneralServiceFormSection.js [hunk 3]
```diff
@@ -92,7 +182,47 @@ class GeneralServiceFormSection extends Component {
...
+        <span>
+        {'You can run Docker containers with both container runtimes. The Universal Container Runtime is better supported in DC/OS. '}
+          <a href={MetadataStore.buildDocsURI('/usage/containerizers/')} target="_blank">
+          More information
+        </a>.
+      </span>
+    );
+
+    return (
+        <div>
```
Comment (nLight) [inline plugins/services/src/js/components/forms/GeneralServiceFormSection.js]: Let's indent this block properly

File: plugins/services/src/js/components/forms/PodContainerServiceFormSection.js [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+              <FieldInput
+                  name={`${path}.artifacts.${index}`}
+                  placeholder="http://"
+                  value={item} />
+              <FieldError>{artifactError}</FieldError>
+            </FormGroup>
+            <FormGroup className="flex flex-item-align-end column-2">
+              <a className="button button-primary-link button-flush"
+                  onClick={this.props.onRemoveItem.bind(this, {value: index, path: `${path}.artifacts`})}>
+                Delete
```
Comment (nLight) [inline plugins/services/src/js/components/forms/PodContainerServiceFormSection.js]: Let's use the new `DeleteRowButton`

File: plugins/services/src/js/components/modals/NewCreateServiceModal.js [hunk 5]
```diff
@@ -251,6 +256,15 @@ class NewServiceFormModal extends Component {
...
 
   getModalContent() {
+    let {
+      isJSONModeActive,
+      serviceConfig,
+      serviceFormActive,
+      serviceJsonActive,
+      servicePickerActive,
+      serviceReviewActive
+    } = this.state;
```
Comment (nLight) [inline plugins/services/src/js/components/modals/NewCreateServiceModal.js]: 👍

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js [hunk 6]
```diff
@@ -212,7 +231,7 @@ class NewCreateServiceModalForm extends Component {
     });
     let patch = batch.reduce(this.props.jsonConfigReducers, {});
 
-    return CreateServiceModalFormUtil.applyPatch(baseConfig, patch);
+    return new CreateServiceModalFormUtil.applyPatch(baseConfig, patch);
```
Comment (nLight) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: why do we need `new` here?

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  let newState = state.slice();
+  const index = path[1];
+  const joinedPath = path.join('.');
+
+  if (joinedPath === 'containers') {
+    switch (type) {
+      case ADD_ITEM:
+        newState.push({name: `container${newState.length + 1}`});
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: Everywhere in the UI we fallback to `container[space]<index>` let's add space here or remove in the UI 💪

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+
+  if (type === SET && joinedPath === `containers.${index}.image`) {
+    newState[index] = Object.assign({}, newState[index], {image: value});
+  }
+
+  return newState;
+};
+
+function containersParser(state) {
+  if (state != null && state.containers != null) {
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: Let's use an early return instead

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        if (resources.mem != null) {
+          memo.push(new Transaction(['containers', index, 'resources', 'mem'], resources.mem));
+        }
+        if (resources.disk != null) {
+          memo.push(new Transaction(['containers', index, 'resources', 'disk'], resources.disk));
+        }
+      }
+
+      //  "privileged": true,
+      // "forcePullImage": true
```
Comment (nLight) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: 🙈

File: plugins/services/src/js/components/forms/GeneralServiceFormSection.js (hunk 1 above)
Comment (Poltergeist) [inline plugins/services/src/js/components/forms/GeneralServiceFormSection.js]: I used the FormGroupContainer here.

File: plugins/services/src/js/components/forms/GeneralServiceFormSection.js (hunk 2 above)
Comment (Poltergeist) [inline plugins/services/src/js/components/forms/GeneralServiceFormSection.js]: `\s` done

File: plugins/services/src/js/components/forms/GeneralServiceFormSection.js (hunk 3 above)
Comment (Poltergeist) [inline plugins/services/src/js/components/forms/GeneralServiceFormSection.js]: done

File: plugins/services/src/js/components/modals/NewCreateServiceModalForm.js (hunk 6 above)
Comment (Poltergeist) [inline plugins/services/src/js/components/modals/NewCreateServiceModalForm.js]: please do not ask ... `removed`

File: plugins/services/src/js/reducers/serviceForm/Containers.js (hunk 7 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: you are right.

File: plugins/services/src/js/reducers/serviceForm/Containers.js (hunk 9 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: you saw nothing.

File: plugins/services/src/js/reducers/serviceForm/Containers.js (hunk 8 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: did this.

File: plugins/services/src/js/components/forms/PodContainerServiceFormSection.js [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+import FieldInput from '../../../../../../src/js/components/form/FieldInput';
+import FieldLabel from '../../../../../../src/js/components/form/FieldLabel';
+import FieldTextarea from '../../../../../../src/js/components/form/FieldTextarea';
+import FormGroup from '../../../../../../src/js/components/form/FormGroup';
+import MetadataStore from '../../../../../../src/js/stores/MetadataStore';
+import Icon from '../../../../../../src/js/components/Icon';
+import {findNestedPropertyInObject} from '../../../../../../src/js/utils/Util';
+
+class PodContainerServiceFormSection extends Component {
+  getAdvancedSettings(data = {}, errors = {}) {
```
Comment (orlandohohmeier) [inline plugins/services/src/js/components/forms/PodContainerServiceFormSection.js]: Please align this section with the single container implementation, which is alway rendering the "Artifact URI" label.

File: plugins/services/src/js/components/forms/PodContainerServiceFormSection.js [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+              value={findNestedPropertyInObject(data, `${path}.resources.disk`)} />
+            <FieldError>{diskErrors}</FieldError>
+          </FormGroup>
+        </div>
+        {artifacts}
+
+        <div>
+          <a className="button button-primary-link button-flush"
+            onClick={this.props.onAddItem.bind(this, {value: 0, path: `${path}.artifacts`})}>
+            Add Artifact
```
Comment (orlandohohmeier) [inline plugins/services/src/js/components/forms/PodContainerServiceFormSection.js]: Please add a "+" at the beginning of the link text.

File: plugins/services/src/js/reducers/serviceForm/Containers.js [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+      case ADD_ITEM:
+        newState[path[1]].artifacts.push('');
+        break;
+      case REMOVE_ITEM:
+        newState[path[1]].artifacts = newState[path[1]].artifacts.filter((item, index) => {
+          return index !== value;
+        });
+        break;
+      case SET:
+        newState[path[1]].artifacts[path[3]] = value;
```
Comment (orlandohohmeier) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: Please adjust this reducer and respective parser as well to output/parse the use the correct artifact scheme.
```JSON
{
        "uri": "http://download.me/file.tgz",
        "extract": true,
        "executable": true,
        "cache": false
}
```
For details please see: https://github.com/mesosphere/marathon/blob/master/docs/docs/rest-api/public/api/v2/types/artifact.raml

**Comment by orlandohohmeier** [comment]
There are a few minor issues around artifacts but the rest looks amazing.  Great job!  ✨ 💯

File: plugins/services/src/js/components/forms/PodContainerServiceFormSection.js (hunk 11 above)
Comment (Poltergeist) [inline plugins/services/src/js/components/forms/PodContainerServiceFormSection.js]: 👍

File: plugins/services/src/js/reducers/serviceForm/Containers.js (hunk 12 above)
Comment (Poltergeist) [inline plugins/services/src/js/reducers/serviceForm/Containers.js]: 👍

**Pull request merged**